File: example_scoring_problems/fibonacci/submissions/accepted/fibonacci.py
#!/usr/bin/env python3
# Correct solution using matrix exponentiation for large n
# For n <= 90, regular iteration works fine
# For n > 90, we use matrix exponentiation with modulo
# @EXPECTED_RESULTS@: CORRECT
# @EXPECTED_SCORE@: 100

MOD = 10**9 + 7


def matrix_mult(A, B, mod):
    return [
        [
            (A[0][0] * B[0][0] + A[0][1] * B[1][0]) % mod,
            (A[0][0] * B[0][1] + A[0][1] * B[1][1]) % mod,
        ],
        [
            (A[1][0] * B[0][0] + A[1][1] * B[1][0]) % mod,
            (A[1][0] * B[0][1] + A[1][1] * B[1][1]) % mod,
        ],
    ]


def matrix_pow(M, n, mod):
    result = [[1, 0], [0, 1]]  # Identity matrix
    while n > 0:
        if n % 2 == 1:
            result = matrix_mult(result, M, mod)
        M = matrix_mult(M, M, mod)
        n //= 2
    return result


def fibonacci(n):
    if n <= 1:
        return n
    if n <= 90:
        # Simple iteration for small n
        a, b = 0, 1
        for _ in range(n - 1):
            a, b = b, (a + b) % MOD
        return b
    else:
        # Matrix exponentiation for large n, with modulo
        M = [[1, 1], [1, 0]]
        result = matrix_pow(M, n, MOD)
        return result[0][1]


n = int(input())
print(fibonacci(n))

File: example_scoring_problems/fibonacci/submissions/accepted/fibonacci.cpp
// @EXPECTED_RESULTS@: CORRECT
// @EXPECTED_SCORE@: 100
#include <iostream>
using namespace std;

const long long MOD = 1e9 + 7;

void matmult(long long A[2][2], long long B[2][2], long long C[2][2]) {
    long long temp[2][2];
    for (int i = 0; i < 2; i++)
        for (int j = 0; j < 2; j++)
            temp[i][j] = (A[i][0] * B[0][j] + A[i][1] * B[1][j]) % MOD;
    for (int i = 0; i < 2; i++)
        for (int j = 0; j < 2; j++)
            C[i][j] = temp[i][j];
}

long long fibonacci(long long n) {
    if (n <= 1) return n;
    long long result[2][2] = {{1, 0}, {0, 1}};
    long long M[2][2] = {{1, 1}, {1, 0}};
    while (n > 0) {
        if (n % 2 == 1) matmult(result, M, result);
        matmult(M, M, M);
        n /= 2;
    }
    return result[0][1];
}

int main() {
    long long n;
    cin >> n;
    cout << fibonacci(n) << endl;
    return 0;
}
